#ism93

#The scripts provided in this repository allows to analyze raw (fastq.qz) data for whole genome sequencing and RNA-seq analyses presented in our paper (under review):

Title: A signaling cohort of IRE1 and proteasome system controls cell fate determination in unresolved proteotoxic stress of the plant endoplasmic reticulum

Software and version information:
"bsa/" folder contains scripts for quality control, mapping, variant calling and bulk segregation analysis. I used the following versions of the programs: FastQC (version 0.11.5), Cutadapt (version 1.8.1), Bowtie (version 2.2.3), Samtools (version 1.8), bcftools (version 1.9.64), SHOREmap (version 3.6). 

"rna_seq/" folder contains scripts for Quality Control, Mapping. We used the following versions of the programs: FastQC (version 0.11.5). Cutadapt (version 1.8.1) Bowtie (version 2.2.4) TopHat (version 2.0.14) Cufflinks (version 1.3.0) R (version 3.4.0) Integrative Genome Browser (version 2.5.0)
